### DISCUSSION TITLE
Use modern python development practices

### DISCUSSION THREAD

**Discussion by Benjamin-Nussbaum** [PublicQuantumNetwork/pqn-stack#5] [General]
The following recommendations are heavily influenced by https://www.stuartellis.name/articles/python-modern-practices/. See the other repositories in this organization for quick reference of how I have already been implementing these concepts.

In particular, we should:

- [Develop for the most recent version of python available](https://www.stuartellis.name/articles/python-modern-practices/#use-the-most-recent-version-of-python-that-you-can)

  - Currently, that is [3.12](https://devguide.python.org/versions/), but future development should ideally target 3.13 after it is released in October.

- [Use a project tool](https://www.stuartellis.name/articles/python-modern-practices/#use-a-project-tool)

  - The other repos currently use [PDM](https://pdm-project.org/en/latest/).

- Lint and format with [Ruff](https://docs.astral.sh/ruff/). See implementation in the other repos in this org.

- [Use TOML for configuration](https://www.stuartellis.name/articles/python-modern-practices/#use-the-toml-format-for-configuration) and data files which must be human-read/writable.
  
  - JSON is fine for passing data when humans don't need to directly interact with the file
  
  - YAML and INI should be avoided

- [Use a `pyproject.toml` file](https://www.stuartellis.name/articles/python-modern-practices/#use-a-pyprojecttoml-file) to store project information and list dependencies.

  - The main thing I've added to the `pyproject.toml`s in the other repos is a slew of lint rules to be used with Ruff.

- [Use the `src/` layout directory structure](https://www.stuartellis.name/articles/python-modern-practices/#create-a-directory-structure-that-uses-the-src-layout)

  - See the [official documentation](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/) for additional information

**Comment by Benjamin-Nussbaum** [comment]
@nunezco2 @marcosfrenkel Let me know what you think. If it looks good, I am happy to take care of this boilerplate setup.

**Reply by marcosfrenkel** [reply]
This all looks great! Feel free to add any of them!

**Comment by Benjamin-Nussbaum** [comment]
Other things to look into:

- [`Protocol`s](https://docs.python.org/3/library/typing.html#typing.Protocol) (alternative to ABCs with some advantages)
- [Pydantic](https://docs.pydantic.dev/latest/) (library for data validation and other nice features)

**Reply by nunezco2** [reply]
Hi Benjamin!

`Pydantic` is a great idea, let's adopt it. I will create a new issue for it.

Regarding `Protocol', I am against introducing it now. We are in a conceptualization phase where code structure will change rapidly and we use it to think about hierarchies of inheritance. Once that stabilizes, we will likely do this.

**Collective development practices.** When producing a pull request, it should address one and only one issue. Otherwise, checking functionality and correctness becomes much harder.

For example, introducing `Protocol` in the future will be a separate issue. I will start creating them per some of your recommendations above.

**Comment by Benjamin-Nussbaum** [comment]
@nunezco2 @marcosfrenkel, 

Re: import sorting

I prefer to have every import on its own line to reduce possible merge conflicts.

However, it's easy enough to change the formatting style in `pyproject.toml` ~to any of these options: https://pycqa.github.io/isort/docs/configuration/multi_line_output_modes.html~ (I guess they're [still working on that...](https://github.com/astral-sh/ruff/issues/2600)) with any of the options described in the Ruff documentation here: https://docs.astral.sh/ruff/settings/#lintisort
